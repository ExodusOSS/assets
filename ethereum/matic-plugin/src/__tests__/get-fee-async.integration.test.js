import { getPrivateSeed, walletTester } from '@exodus/assets-testing'
import { convertForSnapshot } from '@exodus/ethereum-api/src/__tests__/eth-test-utils.js'

import assetPlugin from '../index.js'

const walletAccount = 'exodus_0'

describe('matic async fees for nfts', () => {
  walletTester({
    assetPlugin,
    assetName: 'matic',
    seed: getPrivateSeed(),
    tests: {
      'get fee for nft 1155': async ({ asset, fees: feesModule, assetClientInterface }) => {
        const nft = {
          id: 'polygon:0xbb42bf5aebbbeaef9335d8b497fb2cff2a010eed/1',
          contractAddress: '0xbb42bf5aebbbeaef9335d8b497fb2cff2a010eed',
          tokenId: '1',
          contractType: 'erc1155',
        }
        const fromAddress = '0x464F62159e067a8C231b5c3F9A7144E191750051'
        const toAddress = '0x464F62159e067a8C231b5c3F9A7144E191750051'

        const feeData = await assetClientInterface.getFeeData({ assetName: asset.name })
        const fees = await feesModule.getFees({
          feeData,
          assetName: asset.name,
          walletAccount,
          nft,
          fromAddress,
          toAddress,
        })
        asset.server.stop()
        expect(convertForSnapshot(fees)).toMatchSnapshot()
      },

      'get fee for nft 721': async ({ asset, fees: feesModule, assetClientInterface }) => {
        const nft = {
          id: 'polygon:0xbb42bf5aebbbeaef9335d8b497fb2cff2a010eed/1',
          contractAddress: '0xbb42bf5aebbbeaef9335d8b497fb2cff2a010eed',
          tokenId: '1',
          contractType: 'erc721',
        }
        const fromAddress = '0x464F62159e067a8C231b5c3F9A7144E191750051'
        const toAddress = '0x464F62159e067a8C231b5c3F9A7144E191750051'

        const feeData = await assetClientInterface.getFeeData({ assetName: asset.name })
        const fees = await feesModule.getFees({
          assetName: asset.name,
          feeData,
          nft,
          fromAddress,
          toAddress,
          walletAccount,
        })
        asset.server.stop()
        expect(convertForSnapshot(fees)).toMatchSnapshot()
      },

      'get fee invalid contract broken': async ({ asset, assetClientInterface }) => {
        const nft = {
          id: 'polygon:invalid/1',
          contractAddress: 'invalid',
          tokenId: '1',
        }
        const fromAddress = '0x464F62159e067a8C231b5c3F9A7144E191750051'
        const toAddress = '0x464F62159e067a8C231b5c3F9A7144E191750051'
        const feeData = await assetClientInterface.getFeeData({ assetName: asset.name })

        try {
          await asset.api.getFeeAsync({
            asset,
            feeData,
            nft,
            fromAddress,
            toAddress,
            walletAccount,
          })
          asset.server.stop()
          expect(false).toEqual(true)
        } catch (e) {
          asset.server.stop()
          expect(e.message).toEqual(
            'Cannot fetch gas for matic\n' +
              'ContractType erc1155 nft input 0xf242432a000000000000000000000000464f62159e067a8c231b5c3f9a7144e191750051000000000000000000000000464f62159e067a8c231b5c3f9a7144e1917500510000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000. Bad rpc response: Invalid address\n' +
              'ContractType erc721 nft input 0x42842e0e000000000000000000000000464f62159e067a8c231b5c3f9a7144e191750051000000000000000000000000464f62159e067a8c231b5c3f9a7144e1917500510000000000000000000000000000000000000000000000000000000000000001. Bad rpc response: Invalid address'
          )
        }
      },
    },
  })
})
