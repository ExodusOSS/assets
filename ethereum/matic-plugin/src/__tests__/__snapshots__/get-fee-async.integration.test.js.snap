exports[`matic async fees for nfts get fee for nft 1155 1`] = `
{
  "extraFeeData": {},
  "fee": "0.00325836 MATICNATIVE",
  "gasLimit": 24136,
  "gasPrice": "0.000000135 MATICNATIVE",
  "tipGasPrice": "0.000000035 MATICNATIVE",
  "unsignedTx": {
    "txData": {
      "chainId": 137,
      "transactionBuffer": "02f8f2818980850826299e00851f6ea08600825e4894bb42bf5aebbbeaef9335d8b497fb2cff2a010eed80b8c4f242432a000000000000000000000000464f62159e067a8c231b5c3f9a7144e191750051000000000000000000000000464f62159e067a8c231b5c3f9a7144e1917500510000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000000a00000000000000000000000000000000000000000000000000000000000000000c0808080",
    },
    "txMeta": {
      "amount": "0 MATICNATIVE",
      "assetName": "matic",
      "bumpTxId": undefined,
      "eip1559Enabled": true,
      "fee": "0.00325836 MATICNATIVE",
      "fromAddress": "0x464F62159e067a8C231b5c3F9A7144E191750051",
      "toAddress": "0x464F62159e067a8C231b5c3F9A7144E191750051",
    },
  },
}
`;

exports[`matic async fees for nfts get fee for nft 721 1`] = `
{
  "extraFeeData": {},
  "fee": "0.003051 MATICNATIVE",
  "gasLimit": 22600,
  "gasPrice": "0.000000135 MATICNATIVE",
  "tipGasPrice": "0.000000035 MATICNATIVE",
  "unsignedTx": {
    "txData": {
      "chainId": 137,
      "transactionBuffer": "02f892818980850826299e00851f6ea0860082584894bb42bf5aebbbeaef9335d8b497fb2cff2a010eed80b86442842e0e000000000000000000000000464f62159e067a8c231b5c3f9a7144e191750051000000000000000000000000464f62159e067a8c231b5c3f9a7144e1917500510000000000000000000000000000000000000000000000000000000000000001c0808080",
    },
    "txMeta": {
      "amount": "0 MATICNATIVE",
      "assetName": "matic",
      "bumpTxId": undefined,
      "eip1559Enabled": true,
      "fee": "0.003051 MATICNATIVE",
      "fromAddress": "0x464F62159e067a8C231b5c3F9A7144E191750051",
      "toAddress": "0x464F62159e067a8C231b5c3F9A7144E191750051",
    },
  },
}
`;
